0023-communication-protocol.adoc

```diff
@@ -0,0 +1,383 @@
+In other words, no need to have `type` fields inside the message, the protocol already identifies the type.
+* cheap:
+Opening substreams is very cheap compared to network connections.
+While there is still some inefficiency in terms of opening substreams, a more flexible design seems to be desirable over some added latency.
+Plus, https://github.com/libp2p/specs/pull/205/files[work is underway] to make this more efficient.
+
+=== Communication requirements
+
+In order to design the ideal communication protocol, we have to identify which data we want to transfer and why.
+This section aims to achieve that.
+
+==== Verifying the details of the swap
+
+One of the primary requirements for the communication protocol is for both parties to verify that they talk about the same swap.
+This is crucial in avoiding unnecessary transactions, as long as both parties are willing to actually swap.
+
+==== Sending over protocol-specific data
+
+For HTLC-based protocols like HAN, HErc20 and HALight, Alice generates a secret.
+The communication protocol must provide a way for Alice to share the hash of this secret with Bob.
+We also have to take into account that in the future, we will have swap protocols that depend on other cryptographic elements like signatures.
+
+==== Sharing identities
+
+Atomic swap protocols center around the idea of providing a time window in which the change in ownership of assets is atomic.
+A core aspect of these protocols are identities.
+The communication protocol must provide a way for both parties to share identities with the other party.
+For on-chain (i.e. Layer 1) protocols like HAN and HErc20, this is always a pair of identities (redeem + refund).
+For Layer 2 protocols like HALight, the identity is a single public key: the identifier of the user's lnd instance.
+We have to accommodate for these differences in protocols.
+
+== Recommendation
+
+Based on the research above, this section presents a proposal for a set of libp2p protocols that can be combined to offer the required functionality.
+
+=== Announcing a swap
+
+Protocol-identifier: `/comit/swap/announce/1.0.0`
+
+This protocol includes two messages:
+
+. The announce message
+. The confirmation message
+
+==== Announce-message
+
+[source,json]
+----
+{
+  "swap_digest": "12202c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae"
+}
+----
+
+This message is always sent by Alice.
+See <<Assumptions and goals,assumptions point 3>> for further explanation.
+
+The `swap_digest` field is a https://github.com/multiformats/multihash[multihash-encoded] root hash of a hash list that contains all details of the swap.
+It remains to be defined, how the digest is exactly computed but the idea is the following:
+
+. Take all of the swaps information _known to the COMIT node_ (alpha ledger, alpha asset, protocols, etc) and arrange it in a list
+. Serialize each item into a yet-to-be defined representation
```

Comment:
I find JSON much easier to work with. If we want binary, we can use CBOR.

PS: I am not the only one disliking protobuf :)
https://reasonablypolymorphic.com/blog/protos-are-wrong/

Reply:
I just realized that you commented on the digest computation here.

I wouldn't use JSON for that but neither protobuf. The serialization is going to be very simple, like a single key-value which is then hashed.

An example representation could be:

```rust
format!("alpha_ledger={}", alpha_ledger)
```

and then hashing it.

I don't see a point in using a technology like protobuf here, we don't have any structs or anything which kind of defeats the point of an IDL.

Note: Very likely, we don't want to go with the above example because it is string-encoding dependent but rather define our own binary encoding.